eth_account/typed_transactions/set_code_transaction.py

```diff
@@ -0,0 +1,282 @@
+    Optional,
+    Tuple,
+    cast,
+)
+
+from eth_rlp import (
+    HashableRLP,
+)
+from eth_utils import (
+    keccak,
+)
+from eth_utils.curried import (
+    apply_formatters_to_dict,
+)
+from eth_utils.toolz import (
+    dissoc,
+    merge,
+    partial,
+    pipe,
+)
+from hexbytes import (
+    HexBytes,
+)
+import rlp
+from rlp.sedes import (
+    Binary,
+    CountableList,
+    big_endian_int,
+    binary,
+)
+
+from eth_account._utils.transaction_utils import (
+    transaction_rlp_to_rpc_structure,
+    transaction_rpc_to_rlp_structure,
+)
+from eth_account._utils.validation import (
+    LEGACY_TRANSACTION_VALID_VALUES,
+    is_int_or_prefixed_hexstr,
+    is_rpc_structured_access_list,
+    is_rpc_structured_authorization_list,
+)
+from eth_account.types import (
+    Blobs,
+)
+
+from .access_list_transaction import (
+    access_list_sede_type,
+)
+from .base import (
+    TYPED_TRANSACTION_FORMATTERS,
+    _TypedTransactionImplementation,
+)
+
+
+class Authorization(HashableRLP):
+    magic = 5  # '0x05' 'MAGIC' as per specification EIP-7702
+    fields = (
+        ("chainId", big_endian_int),
+        ("address", Binary.fixed_length(20, allow_empty=False)),
+        ("nonce", big_endian_int),
+    )
+
+    def hash(self) -> HexBytes:
+        """
+        :returns: the hash of the encoded bytestring
+        :rtype: ~hexbytes.main.HexBytes
+        """
+        return HexBytes(
+            pipe(
+                self,
+                rlp.encode,
+                lambda val: bytes([self.__class__.magic]) + val,  # (0x05 || rlp([...]))
+                keccak,
+            )
+        )
+
+
+class SignedAuthorization(HashableRLP):
+    _authority: bytes
+
+    fields = (
+        ("chainId", big_endian_int),
+        ("address", Binary.fixed_length(20, allow_empty=False)),
+        ("nonce", big_endian_int),
+        ("yParity", big_endian_int),
+        ("r", big_endian_int),
+        ("s", big_endian_int),
+    )
+
+
+class SetCodeTransaction(_TypedTransactionImplementation):
+    """
+    Represents a set code transaction as per EIP-7702.
+    """
+
+    transaction_type = 4  # '0x04'
+
+    unsigned_transaction_fields = (
+        ("chainId", big_endian_int),
+        ("nonce", big_endian_int),
+        ("maxPriorityFeePerGas", big_endian_int),
+        ("maxFeePerGas", big_endian_int),
+        ("gas", big_endian_int),
+        ("to", Binary.fixed_length(20, allow_empty=True)),
+        ("value", big_endian_int),
+        ("data", binary),
+        ("accessList", access_list_sede_type),
+        ("authorizationList", CountableList(SignedAuthorization)),
+    )
+
+    signature_fields = (
+        ("v", big_endian_int),
+        ("r", big_endian_int),
+        ("s", big_endian_int),
+    )
+
+    transaction_field_defaults = {
+        "type": b"0x4",
+        "chainId": 0,
+        "nonce": 0,
+        "to": b"",
+        "value": 0,
+        "data": b"",
+        "accessList": [],
+        "authorizationList": [],
```

Comment:
It looks like sometimes we allow a List and sometimes we allow a Tuple for authorizationList, and we should make sure that we're consciously allowing both. I don't have a strong preference for either allowing both or picking one, but just wanted to flag it.